### PR TITLE
Added a "Shopping List" mode feature

### PR DESCRIPTION
**Pull request by williammetcalf**
- Added a "Shopping List" mode, where you can specify items that you want and the scan will inform you of everything you are missing. It also adds a dynamically added "Trash" recipe that identifies items in your inventory that you don't need to accomplish your goal
  - If it isn't clear in the screenshots, the recipe list is filtered to only show the recipes that are in the chain for the specified shopping list
  - The recipe list also includes the "trash" recipe, which selects up to 4 items that aren't in the shopping list tree
PathOfExile_wBUXXeDmJc
PathOfExile_MVR1rlmPUG

- I also just refactored the initial code base to break it out into multiple files, which I think is just easier to maintain
  - The first two commits are 100% the refactor, all commits beyond that are my feature changes
- Fixed misspelled `Treant Horder` > `Treant Horde`
- I noticed the `settings.ini` file wasn't commited to the repo, so I added it to the `.gitignore`

### CODE REVIEW

**Comment by 4rtzel** [comment]
That's a lot of work! [Einhar approves](https://oriath.net/Audio/Dialogue/NPC/BetrayalLeague/Einhar/EF_14_B.ogg)

I'll try to review it as soon as possible. Trying to catch up with real life stuff after providing support for this tool for last 5 days 8 hours a day.

**Comment by meepen** [comment]
Haha, both of us did th same two small things with settings and rename treant :)

**Comment by JustLKS** [comment]
Omg thats amazing! where can I doiwnload it williammetcalf?

**Comment by williammetcalf** [comment]
@4rtzel  alright I added some error handling. TBH I haven't written python in like 7 years I think, and I'm definitely no expert with using TKinter, so idk if I did this especially well. I'm definitely open to suggestions on how to handle errors for this type of thing. 

I'm also sure it would be a bit confusing to a new user to understand exactly what to put in for the recipes, since its case/hypen sensitive

python3 8_VWonfWPyMG

**Comment by williammetcalf** [comment]
> Omg thats amazing! where can I doiwnload it williammetcalf?

For the time being you would need to clone my fork of this repository and setup/run the python code directly. One this PR gets merged the project will need to be re-released

File: src/UIOverlay.py [hunk 1]
```diff
@@ -451,14 +455,21 @@ def _update_shopping_list_mode(self) -> None:
...
-        # TODO: validate it
         shopping_list = list(map(lambda x: x.strip(), self._shopping_list_entry.get().split(",")))
+        if len(shopping_list) == 0 or len(self._shopping_list_entry.get().strip()) == 0:
+            self._update_shopping_list_label("Error: Must enter at least one item")
+            raise ValueError
         for item in shopping_list:
             if item not in self._items_map.items():
-                print("Unknown Recipe:", item)
+                self._update_shopping_list_label('Error: unknown item "{0}"'.format(item))
                 raise ValueError
```
Comment (4rtzel) [inline src/UIOverlay.py:465]: I think you could simple return from this function here instead of raising an exception.

**Comment by 4rtzel** [comment]
> @4rtzel alright I added some error handling. TBH I haven't written python in like 7 years I think, and I'm definitely no expert with using TKinter, so idk if I did this especially well. I'm definitely open to suggestions on how to handle errors for this type of thing.
> 
> I'm also sure it would be a bit confusing to a new user to understand exactly what to put in for the recipes, since its case/hypen sensitive
> 
> python3 8_VWonfWPyMG

I'm not a Python developer either so I can't really tell how good/bad it is :). For me it seems "good enough". Especially giving that this tool is a temporary solution for the UI problems in PoE.

As long as the main feature of your changes is functional, I'll merge them. You can leave the polishing to me.

**Comment by williammetcalf** [comment]
I've been using it myself for a few days and it does seem to work. There's a few more additions I'd like to make but cant be bothered until I see what changes GGG makes.

Ideally I'd like it to show on-screen the items you are missing to make it easier to keep an eye out for them, and it would be nice if it filtered the recipes shown even smarter to not lead you to making duplicates of items you've already made. But I think this is a good baseline for the concept

**Comment by 4rtzel** [comment]
I agree with you about waiting for the solution that GGG will come up with, before doing any more changes. My current plan is to merge your work and release v0.2.0 with keyboard changes.

After that, I'll merge https://github.com/4rtzel/poe-archnemesis-scanner/pull/35 for v0.2.1 release (@meepen changes modify the algorithm quite significantly, so it would be nice if people had a version to fallback to in case of any issues).

**Comment by 4rtzel** [comment]
Played a little bit with the changes. Besides the issues, that I mentioned above (already fixed), everything seems to be working fine.

The code also looks much better now with the split. There are a few stylistic issues but nothing critical. Merging.

**Pull request merged**